Point to the latest JDT and asm

vegegoku opened the pull request:
Update build scripts to point to the latest JDT [3.36.0.v20231115-1055] and the latest asm 9.6

fix gwtproject/gwt#9871
related to gwtproject/gwt#9869

vegegoku committed:
Point to the latest JDT and asm

Update build scripts to point to the latest compatible JDT [3.32.0.v20221108-1853]/[4.26] and the latest asm 9.6

fix gwtproject/gwt#9871
related to gwtproject/gwt#9869

niloc132 commented on .gitignore:
```diff
@@ -6,3 +6,4 @@
 /build/
 # Generated poms
 /maven/poms/**/pom.xml
+/.idea/
```
See the comment at the top of the file and the link on gwtproject.org - maybe the readme should point that out too?

niloc132 commented:
Do you think it would make sense to also update SourceLevel to accept 17 as a value, and pass that to JDT? That should at least get us text blocks and sealed classes "for free" (though of course we still need tests).

I think that makes sense to include in this PR, and give that a quick test so that we know that the new jdt actually does something.

jnehlmeier commented:
This version should also already support Java 21 (except some preview features).

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/890

Most interesting would be record patterns and pattern matching for switch.

niloc132 commented:
@jnehlmeier unfortunately Ahmad has found that builds of JDT past 4.27 require a minimum of Java 17 to compile and run, so if we're dropping just Java 8 for now, we can't update to 4.29. Additionally, there is something funny happening with a new required dependency of the jdt core, and shading it into the gwt-dev.jar seems to prevent classes from being loaded in that jar correctly. 

If you want to push for going all the way to Java 21 now, you could take that up, discuss dropping both Java 8 and 11 on the mailing list, and work up a fix for the new batch dependency? Otherwise I continue to be in favor of a more incremental approach (especially as record/pattern matching is likely going to require more custom codegen to make sense in js).

niloc132 commented on dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java:
```diff
@@ -537,13 +537,20 @@ public void endVisit(BreakStatement x, BlockScope scope) {
 
     @Override
     public void endVisit(CaseStatement x, BlockScope scope) {
+      assert x.isExpr == false : "Switch expressions not yet supported";
```
Let's make this a Precondition instead, so it doesn't get skipped if asserts are disabled? 

(Note that this is feedback on my own original patch, I wrote this in the first draft, intending to do all testing with asserts enabled, but that isn't what a real user would be likely to do.)

niloc132 commented on dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * Tests that {@link GwtAstBuilder} correctly builds the AST for
+ * features introduced in Java 11.
```
```suggestion
 * features introduced in Java 17.
```

niloc132 committed:
Update CI to use 11+, use 21 for artifacts

niloc132 committed:
Update errorprone, remove isJava8 checks

niloc132 committed:
Distribute errorprone suppression, add -noverify

niloc132 committed:
Temporary commit to test nightlies

niloc132 committed:
Disable java21 builds until asm is updated

niloc132 committed:
Roll back noverify, we don't need it with the current build setup

niloc132 committed:
Simplest workaround for https://bugs.openjdk.org/browse/JDK-8323657

niloc132 committed:
Also pass noverify to ci's test.jvmargs

vegegoku committed:
Fix fork CI tools reference

Point CI to the forked tools repository that has java 17 support changes.

fix gwtproject/gwt#9871
related to gwtproject/gwt#9869

niloc132 commented on .github/workflows/full-check.yml:
```diff
@@ -20,7 +20,7 @@ jobs:
     strategy:
       matrix:
-        java-version: [ '8', '11', '17' ]
+        java-version: [ '11', '17' ]
```
Can you re-add 21 in your branch? with the asm update, this should work correctly.

niloc132 commented on .github/workflows/full-check.yml:
```diff
@@ -83,7 +84,7 @@ jobs:
         # Using the same java version as above, set up a settings.xml file
         uses: actions/setup-java@v4
-        if: ${{ github.event_name == 'schedule' && github.repository_owner == 'gwtproject' && matrix.java-version == '17' }}
+        if: ${{ github.event_name == 'schedule' && github.repository_owner == 'gwtproject' && matrix.java-version == '21' }}
```
I should drop these back down to 17 in my branch, but yours should keep 21

niloc132 commented on dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java:
```diff
@@ -537,13 +537,25 @@ public void endVisit(BreakStatement x, BlockScope scope) {
     @Override
     public void endVisit(CaseStatement x, BlockScope scope) {
+      if (x.isExpr) {
+        InternalCompilerException excption =
+            new InternalCompilerException("Switch expressions not yet supported");
+        excption.addNode(new JCaseStatement(makeSourceInfo(x), null));
+        throw excption;
```
```suggestion
        InternalCompilerException exception =
            new InternalCompilerException("Switch expressions not yet supported");
        exception.addNode(new JCaseStatement(makeSourceInfo(x), null));
        throw exception;
```

niloc132 replied:
Should we also have a temporary exception in the visit/endVisit for classes to make sure they aren't records, and a test to confirm?

vegegoku replied:
If we are throwing exception on visit, do we still need to do the same for endVisit?

niloc132 replied:
No, if a visitor throws, it will stop visiting.

I think we need more testing of switch exprs to fail, this doesn't cause an error (since it doesn't have a `case`):
```
    int i = switch(0) {
      default:
        yield 0;
    };
```

Actually I can't get it to fail even with a case:
```
    int i = switch(0) {
      case 1:
        yield 2;
      default:
        yield 7;
    };
```
Maybe we also need a check on the switchstatement? Yeah, confirmed, this only fails using `case ... -> ...` or `default -> ...`.

niloc132 replied:
Still need a test, and failing code for something like this:
```java
    int i = switch(1) {
      case 1:
        yield 2;
      default:
        yield 7;
    };
```

I think you need a visit/endVisit for SwitchExpression to catch this - the error thrown above is only good for the new `case Foo ->` syntax.

niloc132 commented on dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019 Google Inc.
```
2023/2024

niloc132 commented on dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:
```diff
@@ -0,0 +1,126 @@
+ * Tests that {@link GwtAstBuilder} correctly builds the AST for
+ * features introduced in Java 17.
+ */
+public class Java17AstTest extends FullCompileTestBase {
```
Looks like we need one more test class to be complete - the Java17AstTest runs parts of the compiler and makes sure that java can be translated to "gwt java", but we also need a class like Java11Test (and supersource...) to verify that it actually gets compiled to js and works as expected.

niloc132 commented on dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:
```diff
@@ -0,0 +1,126 @@
+          "    case JANUARY, JUNE, JULY -> 3;\n" +
+          "    case FEBRUARY, SEPTEMBER, OCTOBER, NOVEMBER, DECEMBER -> 1;\n" +
+          "    case MARCH, MAY, APRIL, AUGUST -> 2;\n" +
+          "    default -> 0;" +
```
aside: what's the pattern here? J means 3, M and A means 2, everything else is 1?

vegegoku replied:
It is just a random pattern for testing, didnt need to mean anything.

vegegoku committed:
Add js compile tests

Write tests to confirm the code compiles to javascript too, add a test to guard for using records until support is added.

fix gwtproject/gwt#9871
related to gwtproject/gwt#9869

niloc132 reviewed (commented):
I didn't think this through - but I think you're going to need to re-introduce the isJava8 flag as isJava11 so that the Java 17 test won't try to be compiled with Java 11. LMK if you want a hand, this Ant wiring is a pain.

niloc132 commented:
Cherry-pick that last commit to its own PR? Looks like it will need to have at least one line wrapped too, to fix stylecheck.

niloc132 commented on dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java:
```diff
@@ -2752,6 +2764,12 @@ protected void pushBinaryOp(BinaryExpression x, JBinaryOperator op) {
 
     protected boolean visit(TypeDeclaration x) {
+      if (x.isRecord()) {
+        InternalCompilerException excption =
+            new InternalCompilerException("Records not yet supported");
+        excption.addNode(new JClassType(makeSourceInfo(x), null, false, false));
```
```suggestion
        InternalCompilerException exception =
            new InternalCompilerException("Records not yet supported");
        exception.addNode(new JClassType(makeSourceInfo(x), intern(x.name), false, false));
```
Prevents a NPE later in throwing this exception (and fixes a typo).

niloc132 commented on user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java:
```diff
@@ -0,0 +1,80 @@
+ */
+@GwtScriptOnly
+public class Java17Test extends GWTTestCase {
+
```
```suggestion
```

niloc132 commented on user/test/com/google/gwt/dev/jjs/Java17Test.java:
```diff
@@ -0,0 +1,56 @@
+ * NOTE: Make sure this class has the same test methods of its supersourced variant.
+ */
+@DoNotRunWith(Platform.Devel)
+public class Java17Test extends GWTTestCase {
```
I think this is in the wrong package? I think it should be in com/google/gwt/dev/jjs/test instead, that would match the supersource vers

niloc132 commented on .github/workflows/full-check.yml:
```diff
@@ -29,7 +30,8 @@ jobs:
         with:
-          repository: 'gwtproject/tools'
+          repository: 'vegegoku/tools'
+          ref: 'java-17-support'
```
With the tools patch merged, revert this (in both yml files)

niloc132 commented on .github/workflows/full-check.yml:
```diff
@@ -1,4 +1,4 @@
-# Run all tests and builds all aspects of GWT using Java 11 and 17. Runs
+# Run all tests and builds all aspects of GWT using Java 8, 11, and 17. Runs
```
```suggestion
# Run all tests and builds all aspects of GWT using Java 11, 17, and 21. Runs
```

niloc132 commented on dev/build.xml:
```diff
@@ -2,7 +2,7 @@
   <property name="project.tail" value="dev"/>
   <property name="test.args" value="-ea"/>
-  <property name="test.jvmargs" value="-ea"/>
+  <property name="test.jvmargs" value="-ea -agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=8005"/>
```
revert

niloc132 reviewed (commented):
Looks good, just some last minute cleanup. After https://github.com/gwtproject/gwt/pull/9905 is merged let's do one more CI run (to be sure github actions is happy to run on 11/17/21), then merge?
